src/ast.rs

```diff
@@ -1,23 +1,213 @@
-use crate::strtab::Symbol;
+use crate::{lexer::Span, strtab::Symbol};
 
+/// This is the top-level AST node. It stores all class declerations of the
+/// MiniJava program.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Program<'t> {
+    pub span: Span<'t>,
+    pub classes: Vec<ClassDeclaration<'t>>,
+}
+
+/// This AST node stores the Class decleration, which consists of a name and
+/// the members of the class.
+#[derive(Debug, PartialEq, Eq)]
+pub struct ClassDeclaration<'t> {
+    pub span: Span<'t>,
+    pub name: Symbol,
+    pub members: Vec<ClassMember<'t>>,
+}
+
+/// This AST node describes a class member. Variants of class members are
+/// defined in `ClassMemberKind`. Every class member has a type and a name.
+#[derive(Debug, PartialEq, Eq)]
+pub struct ClassMember<'t> {
+    pub span: Span<'t>,
+    pub node: ClassMemberKind<'t>,
+    /// Either the type of a `Field` or the return type of a `Method*`
+    pub ty: Type<'t>,
+    pub name: Symbol,
+    pub is_static: bool,
+}
+
+pub type ParameterList<'t> = Vec<Parameter<'t>>;
+
+/// A class member is either one of
+/// * `Field`: a decleration of a field of a class
+/// * `Method`: a method of a class
+/// * `MainMethod`: a main method, which is a special method that is only
+/// allowed once in a MiniJava Program
+#[derive(Debug, PartialEq, Eq)]
+pub enum ClassMemberKind<'t> {
+    Field,
+    Method(ParameterList<'t>, MethodRest<'t>, Block<'t>),
```

Comment:
`ClassMember` now has an `is_static` field. This can be used identifiy main methods (valid or invalid), as well as the (invalid) static fields. The cue is to move handling of these *invalid* cases to semantical analysis (same goes for new object expression). This was discussed initially [here](https://comprakt.slack.com/archives/CDEADFB6F/p1541010498015500), where the semtiment was negativ, but later I discussed it in RL with @problame and we came to the conclusion to do it anyway, as it makes more sense design-wise (See [this](https://github.com/comprakt/comprakt/pull/20#issuecomment-435273280)).

Reply:
Yeah I just saw that. It also works that way, even it would be cleaner (and more consistent) doing it with an additional enum variant.